### PR TITLE
Remote read first class: query stats middleware

### PR DESCRIPTION
**Pull request by krajorama**
#### What this PR does

Query-frontend: add length, time_since_min_time and time_since_max_time statistics to the query stats logs.
These are generated by the `queryStatsMiddleware` and this PR will enable this middleware for remote read requests.

There are two decisions made by the PR:
- since a remote read request contains multiple queries, we define the overall start, minT and end, maxT times as the minimum and maximum of the individual query times. The assumption is that one remote read contains related queries.
- Mimir ingesters to not take into account the query step hint from remote read requests. To avoid making decisions and updating metrics based on the ignored step hint, we clear the step hint from the request before other middlewares.

#### Which issue(s) this PR fixes or relates to

Related to https://github.com/grafana/mimir-squad/issues/2148 (interna)

#### Checklist

- [X] Tests updated.
- N/A Documentation added.
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- N/A [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/frontend/querymiddleware/stats.go
Comment (pracucci) [inline pkg/frontend/querymiddleware/stats.go:1]: The `queryStatsMiddleware` is also responsible to track `cortex_query_frontend_non_step_aligned_queries_total`. I think we should not track it for remote reads because in Mimir we don't take in account the step at all when executing a remote read request. If we start tracking it for remote reads it will be misleading looking at the `cortex_query_frontend_non_step_aligned_queries_total` metric. It will make us thinking (oh this tenant would benefit a lot by aligning queries) but that may not be true if the metric value is inflated by remote reads.

Comment (krajorama) [reply pkg/frontend/querymiddleware/stats.go:1]: I think the cleanest way then is to set the step to 0 if it's not zero, because then the logic here cannot get out of sync with the ingesters.

**Comment by krajorama** [comment]
Query stats with this patch: query-frontend-1      | ts=2024-06-19T10:05:21.092979449Z caller=handler.go:382 level=info user=anonymous traceID=1b7d4a94a06c1aa2 msg="query stats" component=query-frontend method=POST path=/prometheus/api/v1/read route_name=prometheus_api_v1_read user_agent="mimirtool/unknown (version=unknown, branch=unknown, revision=unknown)" status_code=200 response_time=56.046388ms response_size_bytes=19126 query_wall_time_seconds=0.053434062 fetched_series_count=45 fetched_chunk_bytes=17743 fetched_chunks_count=424 fetched_index_bytes=12612 sharded_queries=0 split_queries=0 estimated_series_count=0 queue_time_seconds=7.096e-05 param_start_0=1718787921000 param_end_0=1718791521000 param_matchers_0="{_ _name_ _=\"up\"}" **length=1h0m0s time_since_min_time=1h0m0.036663817s time_since_max_time=36.663817ms** results_cache_hit_bytes=0 results_cache_miss_bytes=0 status=success

File: pkg/frontend/querymiddleware/stats.go [hunk 1]
```diff
@@ -98,8 +98,12 @@ func (s queryStatsMiddleware) populateQueryDetails(ctx context.Context, req Metr
 	if details == nil {
 		return
 	}
-	details.Start = time.UnixMilli(req.GetStart())
-	details.End = time.UnixMilli(req.GetEnd())
+	if details.Start.IsZero() || details.Start.After(time.UnixMilli(req.GetStart())) {
```
Comment (krajorama) [inline pkg/frontend/querymiddleware/stats.go:101]: Note to reviewers: I'm making the assumption that remote read requests contain related queries so it makes sense to calculate the min of start time and max of end time from a set queries in a remote read request. I'm not sure if this is true, people can easily implement a proxy that batches quires and have nothing in common between them.

Comment (pracucci) [reply pkg/frontend/querymiddleware/stats.go:101]: I think that this is a fair assumption for the Prometheus use case.

File: pkg/frontend/querymiddleware/roundtrip.go [hunk 2]
```diff
@@ -313,6 +313,9 @@ func newQueryMiddlewares(
 	queryStatsMiddleware := newQueryStatsMiddleware(registerer, engine)
 
 	remoteReadMiddleware = append(remoteReadMiddleware,
+		// Track query range statistics. Added first before any subsequent middleware modifies the request.
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/roundtrip.go:316]: I would rephrase this comment given the sanitizer runs before the queryStatsMiddleware.

File: pkg/frontend/querymiddleware/querysharding.go [hunk 3]
```diff
@@ -194,7 +194,19 @@ func newQuery(ctx context.Context, r MetricsQueryRequest, engine *promql.Engine,
 			r.GetQuery(),
 			util.TimeFromMillis(r.GetTime()),
 		)
-
+	case *remoteReadQueryRequest:
+		return engine.NewRangeQuery(
+			ctx,
+			queryable,
+			// We are setting the lookback period to 1 minute and add that amount to
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/querysharding.go:201]: I would better explain why we can't simply set the lookback delta to 0 (because 0 value means used the default). Instead of 1m I would use 1ns to further signal it's just a trick to workaround it.

File: pkg/frontend/querymiddleware/remote_read.go [hunk 4]
```diff
@@ -337,3 +340,43 @@ func cloneRemoteReadQuery(orig *prompb.Query) (*prompb.Query, error) {
...
+func newRemoteReadSanitizerMiddleware(logger log.Logger) MetricsQueryMiddleware {
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &remoteReadSanitizerMiddleware{
+			next:   next,
+			logger: logger,
+		}
+	})
+}
+
+func (r *remoteReadSanitizerMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/remote_read.go:358]: I'm not convinced about the need of the sanitizer. Can't we simply always return 0 from `remoteReadQueryRequest.GetStep()` adding a comment to explain why?

**Review by pracucci — COMMENTED** [review]
Overall LGTM. I just would like to better understand why we need the sanitizer in the first place.

File: pkg/frontend/querymiddleware/remote_read.go (hunk 4 above)
Comment (krajorama) [reply pkg/frontend/querymiddleware/remote_read.go:358]: That was my first reaction as well, but I was afraid of the ingester ending up doing something else.  I'll revert this change and see if I can add a unit test that defines the ingester behavior.

File: pkg/frontend/querymiddleware/roundtrip.go (hunk 2 above)
Comment (krajorama) [reply pkg/frontend/querymiddleware/roundtrip.go:316]: removed sanitizer

File: pkg/frontend/querymiddleware/stats_test.go [hunk 5]
```diff
@@ -97,9 +95,92 @@ func Test_queryStatsMiddleware_Do(t *testing.T) {
...
+			`),
+			expectedQueryDetails: QueryDetails{
+				QuerierStats: &querier_stats.Stats{},
+				Start:        start.Truncate(time.Millisecond),
+				End:          start.Truncate(time.Millisecond),
+				MinT:         start.Truncate(time.Millisecond).Add(-5 * time.Minute),
+				MaxT:         start.Truncate(time.Millisecond),
+			},
+		},
+		"remote read queries without hint": {
```
[the remark this answers is not in the document]
Comment (krajorama) [reply pkg/frontend/querymiddleware/stats_test.go:129]: added

**Review by pracucci — APPROVED** [review]

File: pkg/frontend/querymiddleware/stats.go
Comment (krajorama) [reply pkg/frontend/querymiddleware/stats.go:1]: we decided to ignore hints instead and just return value without taking them into account
